### DISCUSSION TITLE
Could not download averaged_CT-EFT-20.pt

### DISCUSSION THREAD

**Discussion by stefanes1** [sgrvinod/chess-transformers#7] [Q&A] [answered]
I was just experimenting, trying to play vs the model, out of curiosity. I could not access this file "averaged_CT-EFT-20.pt". Was told: "You may still be able to access the file from the browser:" https://drive.google.com/uc?id=1OHtg336ujlOjp5Kp0KjE1fAPF74aZpZD but could not download. Thank you.

**Comment by sgrvinod** [comment, marked as answer]
Hi @stefanes1, thank you for bringing this to my attention. It turns out Google Drive was rate limiting access to the file. It's my fault: Google Drive is not really meant for sharing files with many people, I suppose.

I've now moved model checkpoints and other stuff to Microsoft Azure Storage, and downloading should work now. You'll need to update your local copy to the latest version of this repository (0.2.1). 

You can either download automatically when you attempt to load the model, as shown in [human_play.ipynb](https://github.com/sgrvinod/chess-transformers/blob/main/chess_transformers/play/human_play.ipynb), or download it manually using the [link on the README](https://github.com/sgrvinod/chess-transformers/tree/main?tab=readme-ov-file#ct-eft-20). 

If you download it manually for any reason, save it in a folder called "CT-EFT-20" in a folder called "checkpoints" in "./chess_transformers": therefore, as "./chess_transformers/checkpoints/CT-EFT-20/averaged_CT-EFT-20.pt".

If anything else does not work, let me know. I'd also be curious to hear if you have any impressions after playing. I think you might be the first person to try this repository :)

**Comment by stefanes1** [comment]
Thank you. I hope you are well.

When I run my load_model(CONFIG) now I get this error:

  File "C:\Users\Stefan\Documents\Stef\Python code\Chess Transformers\chess-transformers\chess_transformers\play\utils.py", line 66, in load_model
    gdown.download(
  File "C:\Python311\Lib\site-packages\gdown\download.py", line 259, in download
    filename_from_url = m.groups()[0]
                        ^^^^^^^^
AttributeError: 'NoneType' object has no attribute 'groups'

Do I need to remove gdown dependencies? I saw something in the code there about that? Thanks

**Comment by stefanes1** [comment]
Hi there

Thanks for help. I downloaded manually, saved the file and reran the code. Now getting this runtime error related to CUDA:

Traceback (most recent call last):
  File "C:\Users\Stefan\Documents\Stef\Python code\Chess Transformers\chess-transformers\human_playv2.py", line 12, in <module>
    model = load_model(CONFIG)
            ^^^^^^^^^^^^^^^^^^
  File "C:\Users\Stefan\Documents\Stef\Python code\Chess Transformers\chess-transformers\chess_transformers\play\utils.py", line 71, in load_model
    checkpoint = torch.load(str(checkpoint_path))
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Python311\Lib\site-packages\torch\serialization.py", line 1014, in load
    return _load(opened_zipfile,
           ^^^^^^^^^^^^^^^^^^^^^
  File "C:\Python311\Lib\site-packages\torch\serialization.py", line 1422, in _load
    result = unpickler.load()
             ^^^^^^^^^^^^^^^^
  File "C:\Python311\Lib\site-packages\torch\serialization.py", line 1392, in persistent_load
    typed_storage = load_tensor(dtype, nbytes, key, _maybe_decode_ascii(location))
                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Python311\Lib\site-packages\torch\serialization.py", line 1366, in load_tensor
    wrap_storage=restore_location(storage, location),
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Python311\Lib\site-packages\torch\serialization.py", line 381, in default_restore_location
    result = fn(storage, location)
             ^^^^^^^^^^^^^^^^^^^^^
  File "C:\Python311\Lib\site-packages\torch\serialization.py", line 274, in _cuda_deserialize
    device = validate_cuda_device(location)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Python311\Lib\site-packages\torch\serialization.py", line 258, in validate_cuda_device
    raise RuntimeError('Attempting to deserialize object on a CUDA '
RuntimeError: Attempting to deserialize object on a CUDA device but torch.cuda.is_available() is False. If you are running on a CPU-only machine, please use torch.load with map_location=torch.device('cpu') to map your storages to the CPU.

Is this part of Pytorch package? Thanks

**Reply by sgrvinod** [reply]
Hi @stefanes1, sorry for the late reply. The problem you have with `gdown` should not happen after you update to the latest version, since `gdown` is not used anymore. The checkpoint should automatically download from Azure cloud.

About this new error, it's because (I'm guessing) you don't have a GPU. In this case, [the following line in this file](https://github.com/sgrvinod/chess-transformers/blob/1dcf11085bd6c82dbac59cd4f90b102e4c2f499f/chess_transformers/play/utils.py#L117)

```
    checkpoint = torch.load(str(checkpoint_path))
```

should be changed to:

```
    checkpoint = torch.load(str(checkpoint_path), map_location=torch.device('cpu'))
```

You might need to add `import torch` at the beginning of the file.

Sorry about this, I'll fix this in the next update.

**Reply by sgrvinod** [reply]
Actually, [there's already a device variable called `DEVICE`](https://github.com/sgrvinod/chess-transformers/blob/1dcf11085bd6c82dbac59cd4f90b102e4c2f499f/chess_transformers/play/utils.py#L24) in the file. You can just do `map_location=DEVICE`.

If you're still facing any problem, please let me know.
